tss/node/signer/keygen.go

```diff
@@ -61,3 +75,91 @@ func (p *Processor) Keygen() {
 		}
 	}()
 }
+
+func (p *Processor) setGroupPublicKey(localKey, poolPubkey string) error {
+	p.logger.Info().Msg("connecting to layer one")
+	if err := ensureConnection(p.l1Client); err != nil {
+		p.logger.Err(err).Msg("Unable to connect to layer one")
+		return err
+	}
+	if len(p.tssGroupManagerAddress) == 0 {
+		p.logger.Error().Msg("tss group manager address is empty ")
+		return errors.New("tss group manager address is empty")
+	}
+	address := ethc.HexToAddress(p.tssGroupManagerAddress)
+	chainId, err := p.l1Client.ChainID(p.ctx)
+	if err != nil {
+		return err
+	}
+
+	opts, err := bind.NewKeyedTransactorWithChainID(p.privateKey, chainId)
+	if opts.Context == nil {
+		opts.Context = context.Background()
+	}
+	opts.NoSend = true
+	contract, err := tgm.NewTssGroupManager(address, p.l1Client)
+	if err != nil {
+		return err
+	}
+	gasPrice, err := p.l1Client.SuggestGasPrice(context.Background())
+	if err != nil {
+		p.logger.Err(err).Msg("cannot fetch gas price")
+		return err
+	}
+	opts.GasPrice = gasPrice
+	tx, err := contract.SetGroupPublicKey(opts, []byte(localKey), []byte(poolPubkey))
+	if err != nil {
+		return err
+	}
+	if err := p.l1Client.SendTransaction(p.ctx, tx); err != nil {
+		return err
+	}
+
+	receipt, err := waitForReceipt(p.l1Client, tx)
+	if err != nil {
+		return err
+	}
+	p.logger.Info().Msgf("tss group tss transaction confirmed , hash %s, gas-used %s,block number  %s", tx.Hash().Hex(), receipt.GasUsed, receipt.BlockNumber)
+	return nil
+
+}
+
+//Ensure we can actually connect l1
+func ensureConnection(client *ethclient.Client) error {
+	t := time.NewTicker(1 * time.Second)
+	retries := 0
+	defer t.Stop()
+	for ; true; <-t.C {
+		_, err := client.ChainID(context.Background())
+		if err == nil {
+			break
+		} else {
+			retries += 1
+			if retries > 90 {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Wait for the receipt by polling the backend
+func waitForReceipt(backend *ethclient.Client, tx *etht.Transaction) (*etht.Receipt, error) {
```

Comment:
better set a timeout for the receipt waiting

Reply:
fixed